impeller/renderer/backend/vulkan/swapchain/ahb/ahb_swapchain_impl_vk.cc

```diff
@@ -0,0 +1,222 @@
+    std::weak_ptr<android::SurfaceControl> surface_control,
+    const ISize& size,
+    bool enable_msaa)
+    : surface_control_(std::move(surface_control)),
+      pending_presents_(std::make_shared<fml::Semaphore>(kMaxPendingPresents)) {
+  desc_ = android::HardwareBufferDescriptor::MakeForSwapchainImage(size);
+  pool_ = std::make_shared<AHBTexturePoolVK>(context, desc_);
+  if (!pool_->IsValid()) {
+    return;
+  }
+  transients_ = std::make_shared<SwapchainTransientsVK>(
+      context, ToSwapchainTextureDescriptor(desc_), enable_msaa);
+  is_valid_ = true;
+}
+
+AHBSwapchainImplVK::~AHBSwapchainImplVK() = default;
+
+const ISize& AHBSwapchainImplVK::GetSize() const {
+  return desc_.size;
+}
+
+bool AHBSwapchainImplVK::IsValid() const {
+  return is_valid_;
+}
+
+const android::HardwareBufferDescriptor& AHBSwapchainImplVK::GetDescriptor()
+    const {
+  return desc_;
+}
+
+std::unique_ptr<Surface> AHBSwapchainImplVK::AcquireNextDrawable() {
+  {
+    TRACE_EVENT0("impeller", "CompositorPendingWait");
+    if (!pending_presents_->Wait()) {
+      return nullptr;
+    }
+  }
+
+  AutoSemaSignaler auto_sema_signaler =
+      std::make_shared<fml::ScopedCleanupClosure>(
+          [sema = pending_presents_]() { sema->Signal(); });
+
+  if (!is_valid_) {
+    return nullptr;
+  }
+
+  auto texture = pool_->Pop();
+
+  if (!texture) {
+    VALIDATION_LOG << "Could not create AHB texture source.";
+    return nullptr;
+  }
+
+  auto surface = SurfaceVK::WrapSwapchainImage(
+      transients_, texture,
+      [signaler = auto_sema_signaler, weak = weak_from_this(), texture]() {
+        auto thiz = weak.lock();
+        if (!thiz) {
+          VALIDATION_LOG << "Swapchain died before image could be presented.";
+          return false;
+        }
+        return thiz->Present(signaler, texture);
+      });
+
+  if (!surface) {
+    return nullptr;
+  }
+
+  return surface;
+}
+
+bool AHBSwapchainImplVK::Present(
+    const AutoSemaSignaler& signaler,
+    const std::shared_ptr<AHBTextureSourceVK>& texture) {
+  auto control = surface_control_.lock();
+  if (!control || !control->IsValid()) {
+    VALIDATION_LOG << "Surface control died before swapchain image could be "
+                      "presented.";
+    return false;
+  }
+
+  if (!texture) {
+    return false;
+  }
+
+  auto fence = SubmitCompletionSignal(texture);
+
+  if (!fence) {
+    VALIDATION_LOG << "Could not submit completion signal.";
+    return false;
+  }
+
+  android::SurfaceTransaction transaction;
+  if (!transaction.SetContents(control.get(),               //
+                               texture->GetBackingStore(),  //
+                               fence->CreateFD()            //
+                               )) {
+    VALIDATION_LOG << "Could not set swapchain image contents on the surface "
+                      "control.";
+    return false;
+  }
+  return transaction.Apply([signaler, texture, weak = weak_from_this()]() {
```

Comment:
I'm not sure what thread this runs on, but for hybrid composition today there are some assumptions about flutters presentation, we may need to conditionally block on the presenting thread until we reach some point in presentation, either here or in the worse case later on when it says we're done with the texture.


See https://github.com/flutter/engine/pull/44881 for how I did this originally. Though this code was backed out later so the android code would need to be redone